packages/util-dns/src/NodeDnsLookupHostResolver.ts

```diff
@@ -20,20 +23,144 @@ const DNS_LOOKUP_OPTIONS = {
+    this.cache = config?.cache !== undefined ? config.cache : NodeDnsLookupHostResolver.DEFAULT_CACHE_PROVIDER();
+    this.nodeDnsLookup =
+      config?.nodeDnsLookup !== undefined ? config.nodeDnsLookup : NodeDnsLookupHostResolver.DEFAULT_NODE_DNS_LOOKUP;
+  }
+
+  /**
+   * Resolves the address(es) for {@link HostResolverArguments} and returns a
+   * list of addresses with (most likely) two addresses, one {@link HostAddressType.AAAA}
+   * and one {@link HostAddressType.A}. Calls to this function will likely alter
+   * the cache so that if there's multiple addresses, a different set will be
+   * returned on the next call.
+   * In the case of multi-answer, still only a maximum of two records should be
+   * returned.
+   * @param args arguments with host name query addresses for
+   * @returns promise with a list of {@link HostAddress}
+   * @see https://github.com/awslabs/aws-c-io/blob/f2ff573c191e1c4ea0248af5c08161356be3bc78/source/host_resolver.c#L964
+   */
+  public async resolveAddress(args: HostResolverArguments): Promise<HostAddress[]> {
+    const possibleHostEntry = this.cache.get(args.hostName);
+    const newNextTimestampToUpdateMs = Date.now() + this.ttlMs;
+
+    // If not in cache, create a new entry.
+    // If this fails, then fail resolution.
+    if (possibleHostEntry === undefined) {
+      const addresses = await this.nodeDnsLookupResolveAddress(args);
+      this.cache.set(args, addresses, newNextTimestampToUpdateMs);
+    }
+
+    const hostEntry = this.cache.get(args.hostName)!;
+
+    // Update existing cache entries every once in a while.
+    // Skips if it was the initial host resolution.
+    if (possibleHostEntry !== undefined && Date.now() >= hostEntry.nextTimestampToUpdateMs) {
+      try {
+        const addresses = await this.nodeDnsLookupResolveAddress(args);
+        hostEntry.updateRecords(addresses, newNextTimestampToUpdateMs);
+      } catch (error) {
+        console.error(`Could not update DNS address cache for "${args.hostName}": ${error}`);
+      }
+    }
+
+    hostEntry.processRecords();
+
+    const result: HostAddress[] = [];
+    if (hostEntry.aRecords.length > 0) {
+      result.push(hostEntry.aRecords.cycle());
+    }
+    if (hostEntry.aaaaRecords.length > 0) {
+      result.push(hostEntry.aaaaRecords.cycle());
+    }
+    if (result.length === 0) {
+      throw new Error(`Could not resolve addresses for "${args.hostName}"`);
+    }
+    return result;
+  }
+
   /**
    * Node.js resolveAddress() using the dns.lookup() APIs.
    * @see https://nodejs.org/api/dns.html#dnspromiseslookuphostname-options
    * @param args arguments with host name query addresses for
    * @returns promise with a list of {@link HostAddress}
    */
-  async resolveAddress(args: HostResolverArguments): Promise<HostAddress[]> {
+  private async nodeDnsLookupResolveAddress(args: HostResolverArguments): Promise<HostAddress[]> {
```

Comment:
private methods should go below public methods

Reply:
Resolved in https://github.com/aws/aws-sdk-js-v3/pull/4433/commits/a3eab77afad8781e943b9c0d9c5166d5fb02febb and https://github.com/aws/aws-sdk-js-v3/pull/4433/commits/2cb4adfcb94dc1ad13ba075055e430d1aa50afd7